x/mongo/driver/topology/server.go

```diff
@@ -294,27 +285,61 @@ func (s *Server) RequestImmediateCheck() {
 	}
 }
 
+// ProcessHandshakeError handles connection errors during the handshake.
+func (s *Server) ProcessHandshakeError(err error, conn driver.Connection) {
+	//ignore stale error
+	if conn.Stale() {
+		return
+	}
+
+	wrappedConnErr := unwrapConnectionError(err)
+	if wrappedConnErr == nil {
+		return
+	}
+
+	// Since the only kind of ConnectionError we receive from pool.Get will be an initialization
+	// error, we should set the description.Server appropriately.
+	s.updateDescription(description.NewServerFromError(s.address, wrappedConnErr, s.Description().TopologyVersion))
+	s.pool.clear()
+}
+
 // ProcessError handles SDAM error handling and implements driver.ErrorProcessor.
-func (s *Server) ProcessError(err error) {
-	desc := s.Description()
+func (s *Server) ProcessError(err error, conn driver.Connection) {
+	//ignore stale error
+	if conn.Stale() {
+		return
+	}
 	// Invalidate server description if not master or node recovering error occurs.
 	// These errors can be reported as a command error or a write concern error.
+	desc := conn.Description()
 	if cerr, ok := err.(driver.Error); ok && (cerr.NodeIsRecovering() || cerr.NotMaster()) {
+		// ignore stale error
+		if description.CompareTopologyVersion(desc.TopologyVersion, cerr.TopologyVersion) >= 0 {
+			return
+		}
+
 		// updates description to unknown
-		s.updateDescription(description.NewServerFromError(s.address, err))
+		s.updateDescription(description.NewServerFromError(s.address, err, cerr.TopologyVersion))
+		s.RequestImmediateCheck()
+
 		// If the node is shutting down or is older than 4.2, we synchronously clear the pool
```

Comment:
I pointed this out in Slack, but leaving a comment here for completeness: The pseudocode always requests an immediate check, so this call should be right below the `updateDescription` one.

Reply:
This was corrected while I was reviewing. Resolving.